sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala

```diff
@@ -375,7 +377,11 @@ class TungstenAggregationIterator(
         val buffer: UnsafeRow = hashMap.getAggregationBufferFromUnsafeRow(groupingKey)
         if (buffer == null) {
           // buffer == null means that we could not allocate more memory.
```

Comment:
The root cause, we will switch to sort-based aggregation only we don't have enough memory, however, the first step of sort-based aggregation is spill, which also requires more memory to be allocated.